server/src/main/java/org/apache/iotdb/db/query/executor/PreviousFillExecutor.java

```diff
@@ -0,0 +1,208 @@
+    }
+    if (containedByTimeFilter(pageStatistics)) {
+      return constructLastPair(
+          pageStatistics.getEndTime(), pageStatistics.getLastValue(), dataType);
+    } else {
+      BatchData batchData = pageReader.getAllSatisfiedPageData();
+      return batchData.getLastPairBeforeOrEqualTimestamp(queryTime);
+    }
+  }
+
+  private boolean shouldUpdate(long time, long version, long newTime, long newVersion) {
+    return time < newTime || (time == newTime && version < newVersion);
+  }
+
+  private List<IPageReader> unpackChunkReaderToPageReaderList(ChunkMetaData metaData) throws IOException {
+    if (metaData == null) {
+      throw new IOException("Can't init null chunkMeta");
+    }
+    IChunkReader chunkReader;
+    IChunkLoader chunkLoader = metaData.getChunkLoader();
+    if (chunkLoader instanceof MemChunkLoader) {
+      MemChunkLoader memChunkLoader = (MemChunkLoader) chunkLoader;
+      chunkReader = new MemChunkReader(memChunkLoader.getChunk(), timeFilter);
+    } else {
+      Chunk chunk = chunkLoader.getChunk(metaData);
+      chunkReader = new ChunkReader(chunk, timeFilter);
+      chunkReader.hasNextSatisfiedPage();
+    }
+    return chunkReader.getPageReaderList();
+  }
+
+  private PriorityQueue<TsFileResource> sortUnSeqFileResourcesInDecendingOrder(
+      List<TsFileResource> tsFileResources) {
+    PriorityQueue<TsFileResource> unseqTsFilesSet =
+        new PriorityQueue<>(
+            (o1, o2) -> {
+              Map<String, Long> startTimeMap = o1.getEndTimeMap();
+              Long minTimeOfO1 = startTimeMap.get(seriesPath.getDevice());
+              Map<String, Long> startTimeMap2 = o2.getEndTimeMap();
+              Long minTimeOfO2 = startTimeMap2.get(seriesPath.getDevice());
+
+              return Long.compare(minTimeOfO2, minTimeOfO1);
+            });
+    unseqTsFilesSet.addAll(tsFileResources);
+    return unseqTsFilesSet;
+  }
+
+  /**
+   * unpack all overlapped seq/unseq files and find the first chunk metadata
+   */
+  private void UnpackAllOverlappedFilesToChunkMetadatas() throws IOException {
+    for (int index = seqFileResource.size() - 1; index >= 0; index--) {
+      TsFileResource resource = seqFileResource.remove(index);
+      List<ChunkMetaData> chunkMetadata = FileLoaderUtils.loadChunkMetadataFromTsFileResource(
+          resource, seriesPath, context);
+      if (!chunkMetadata.isEmpty()) {
+        for (int i = chunkMetadata.size() - 1; i >= 0; i--) {
+          if (chunkMetadata.get(i).getStartTime() <= queryTime) {
+            lastSeqChunkMetaData = chunkMetadata.get(chunkMetadata.size() - 1);
+            chunkMetadatas.add(lastSeqChunkMetaData);
+            break;
+          }
+        }
+        break;
+      }
+    }
+
+    while (!unseqFileResource.isEmpty()
+        && (lastSeqChunkMetaData == null || (lastSeqChunkMetaData.getStartTime()
+        <= unseqFileResource.peek().getEndTimeMap().get(seriesPath.getDevice())))) {
+      chunkMetadatas.addAll(
```

Comment:
Suppose the queried time is 100.
seq files: [1,10], [11,20], [200, 300]
unseq files: [30, 40], [41, 50], [51, 60]

It's better not to find the lastSeqChunkMetadata but the last chunkmetadata between seq and unseq.

Reply:
modified